### PR TITLE
Upgrade to oras v0.9.0

### PR DESCRIPTION
**Pull request by jdolitsky**
Please see security advisory: https://github.com/deislabs/oras/security/advisories/GHSA-g5v4-5x39-vwhx

### CODE REVIEW

**Comment by dtrudg** [comment]
Hello @jdolitsky - thank you for submitting this PR to us after the oras 0.9.0 release. 

While we will update our dependency, we have looked into the GHSA linked and do not believe that Singularity is vulnerable to the security issue.

Singularity both:

1) Limits the media types it will accept to SIF, such that under normal circumstances it would never pull a tar artifact.

https://github.com/hpcng/singularity/blob/920a6a451b86a60331c34ca9b88e5ec7b8387b3a/internal/pkg/client/oras/oras.go#L105

2) Registers a PullBaseHandler that examines each descriptor, and explicitly rejects any with the `AnnotationUnpack` set - which governs extraction of a tar by oras. This protects against maliciously submitting a tar, while labelling it as a SIF.

https://github.com/hpcng/singularity/blob/920a6a451b86a60331c34ca9b88e5ec7b8387b3a/internal/pkg/client/oras/oras.go#L111

I can push a tar artefact, incorrectly labelled as a SIF:

```
12:04 PM $ ./oras push localhost:5000/tar-labelled-as-sif:v2  ./docs/:application/vnd.sylabs.sif.layer.v1.sif
Uploading 4b5f93c50d17 docs
Pushed localhost:5000/tar-labelled-as-sif:v2
Digest: sha256:53f84d3dcddaf141d92fe54cd8302c703623bdb87b3c6afaa9ef6c04222084b5
```

Singularity will not retrieve it:

```
12:20 PM $ singularity pull oras://localhost:5000/tar-labelled-as-sif:v2
INFO:    Downloading oras image
FATAL:   While pulling image from oci registry: error fetching image to cache: unable to Download Image: unable to pull from registry: descriptor is of a bundled directory, not a SIF image
```

The registry logs confirm Singularity only fetched the manifest, and never the tar file - so there is no danger of a malicious tar being extracted:

```
12:20 PM $ docker logs -n 0 -f 6062fc079c86
172.17.0.1 - - [22/Jan/2021:18:20:54 +0000] "HEAD /v2/tar-labelled-as-sif/manifests/v2 HTTP/1.1" 200 533 "" "containerd/1.4.3+unknown"
time="2021-01-22T18:20:54.418117903Z" level=info msg="response completed" go.version=go1.11.2 http.request.host="localhost:5000" http.request.id=54163381-504e-447d-a5db-ddf53d114da2 http.request.method=HEAD http.request.remoteaddr="172.17.0.1:48952" http.request.uri="/v2/tar-labelled-as-sif/manifests/v2" http.request.useragent="containerd/1.4.3+unknown" http.response.contenttype="application/vnd.oci.image.manifest.v1+json" http.response.duration=3.986446ms http.response.status=200 http.response.written=533
time="2021-01-22T18:20:54.422219314Z" level=info msg="response completed" go.version=go1.11.2 http.request.host="localhost:5000" http.request.id=261833ed-3b4b-4aba-9b97-47bce2867558 http.request.method=GET http.request.remoteaddr="172.17.0.1:48952" http.request.uri="/v2/tar-labelled-as-sif/manifests/sha256:53f84d3dcddaf141d92fe54cd8302c703623bdb87b3c6afaa9ef6c04222084b5" http.request.useragent="containerd/1.4.3+unknown" http.response.contenttype="application/vnd.oci.image.manifest.v1+json" http.response.duration=2.70839ms http.response.status=200 http.response.written=533
172.17.0.1 - - [22/Jan/2021:18:20:54 +0000] "GET /v2/tar-labelled-as-sif/manifests/sha256:53f84d3dcddaf141d92fe54cd8302c703623bdb87b3c6afaa9ef6c04222084b5 HTTP/1.1" 200 533 "" "containerd/1.4.3+unknown"
172.17.0.1 - - [22/Jan/2021:18:20:54 +0000] "HEAD /v2/tar-labelled-as-sif/manifests/v2 HTTP/1.1" 200 533 "" "containerd/1.4.3+unknown"
time="2021-01-22T18:20:54.428360704Z" level=info msg="response completed" go.version=go1.11.2 http.request.host="localhost:5000" http.request.id=58b21d51-ea99-4a77-a186-e10b1f3ed9fc http.request.method=HEAD http.request.remoteaddr="172.17.0.1:48952" http.request.uri="/v2/tar-labelled-as-sif/manifests/v2" http.request.useragent="containerd/1.4.3+unknown" http.response.contenttype="application/vnd.oci.image.manifest.v1+json" http.response.duration=4.115042ms http.response.status=200 http.response.written=533
172.17.0.1 - - [22/Jan/2021:18:20:54 +0000] "GET /v2/tar-labelled-as-sif/manifests/sha256:53f84d3dcddaf141d92fe54cd8302c703623bdb87b3c6afaa9ef6c04222084b5 HTTP/1.1" 200 533 "" "containerd/1.4.3+unknown"
time="2021-01-22T18:20:54.433107551Z" level=info msg="response completed" go.version=go1.11.2 http.request.host="localhost:5000" http.request.id=d0082b66-65f2-4df3-b0fc-69bcb6f67c6b http.request.method=GET http.request.remoteaddr="172.17.0.1:48952" http.request.uri="/v2/tar-labelled-as-sif/manifests/sha256:53f84d3dcddaf141d92fe54cd8302c703623bdb87b3c6afaa9ef6c04222084b5" http.request.useragent="containerd/1.4.3+unknown" http.response.contenttype="application/vnd.oci.image.manifest.v1+json" http.response.duration=2.389904ms http.response.status=200 http.response.written=533

```

If you believe anything in our analysis is in error, we would be grateful if you could contact us via security@sylabs.io

CC'ing @tri-adam and @ikaneshiro for info.

**Comment by jdolitsky** [comment]
I believe you are correct. Singularity is likely not vulnerable.

I'd still recommend the upgrade since oras v0.8.1 is listed as a dependency in go.mod, and may show up in some static analysis once an official CVE is issued.

**Comment by ikaneshiro** [comment]
My testing has confirmed @dtrudg's and @tri-adam's findings above. 

Just to cover everything I could think of, I also tested with an image where `"io.deis.oras.content.unpack" = "false"` and this does get past our mediaType handler, and allows the blob to be pulled locally...

```
docker logs -f 733111f9d3dfeda6629c650293e374ad115176820aa35519d022aa00b5ffc6f9
...
172.17.0.1 - - [22/Jan/2021:19:23:49 +0000] "GET /v2/tar-labelled-as-sif/manifests/sha256:e5a09c9e88f57f20f73813d1537c7de1ff277c862dec49bef72dc7f4fafcb8ae HTTP/1.1" 200 534 "" "containerd/1.4.3+unknown"
time="2021-01-22T19:23:49.412314925Z" level=info msg="response completed" go.version=go1.11.2 http.request.host="localhost:5000" http.request.id=11a5ba1c-668d-4e40-a03d-7f3e2cd0e77a http.request.method=GET http.request.remoteaddr="172.17.0.1:39196" http.request.uri="/v2/tar-labelled-as-sif/blobs/sha256:2ff8d26edbe51d4e73bf07f31211722313992d218d884e3dc5636a02ee23424d" http.request.useragent="containerd/1.4.3+unknown" http.response.contenttype="application/octet-stream" http.response.duration=2.29407ms http.response.status=200 http.response.written=123 
172.17.0.1 - - [22/Jan/2021:19:23:49 +0000] "GET /v2/tar-labelled-as-sif/blobs/sha256:2ff8d26edbe51d4e73bf07f31211722313992d218d884e3dc5636a02ee23424d HTTP/1.1" 200 123 "" "containerd/1.4.3+unknown"
```

But, it does not trigger any extraction behavior and is immediately removed once the blob fails the SIF format inspection:
```
$ singularity pull oras://localhost:5000/tar-labelled-as-sif:v2
INFO:    Downloading oras image
FATAL:   While pulling image from oci registry: error fetching image to cache: unable to Download Image: could not open image /home/ian/.singularity/cache/oras/tmp_144634409 for verification: image format not recognized
```

@jdolitsky Thanks for the PR. We definitely want to update this dependency either way!

File: go.mod [hunk 1]
```diff
@@ -50,14 +50,17 @@ require (
...
 	gopkg.in/yaml.v2 v2.4.0
 	gotest.tools/v3 v3.0.3
 	mvdan.cc/sh/v3 v3.2.1
-	rsc.io/letsencrypt v0.0.3 // indirect
 )
 
-replace golang.org/x/crypto => github.com/sylabs/golang-x-crypto v0.0.0-20181006204705-4bce89e8e9a9
+replace (
+	github.com/docker/distribution => github.com/docker/distribution v0.0.0-20191216044856-a8371794149d
+	github.com/docker/docker => github.com/moby/moby v17.12.0-ce-rc1.0.20200618181300-9dc6525e6118+incompatible
```
Comment (ikaneshiro) [inline go.mod:65]: @jdolitsky Wrt upgrading this go module, the need for these replacements is curious to me. 

Is there a reason why these changes (https://github.com/deislabs/oras/pull/209/files#diff-33ef32bf6c23acb95f5902d7097b7a1d5128ca061167ec0716715b0b9eeaa5f6L16) needed to be made in the upstream module to move those dependencies to invalid versions?

**Review by ikaneshiro — CHANGES_REQUESTED** [review]

File: go.mod (hunk 1 above)
Comment (jdolitsky) [reply go.mod:65]: It was the result of running `go mod tidy` - if you look at go.sum you'll notice same versions pinned. If you're able to help stabilize these in oras, please submit a PR and we will re-release

Comment (ikaneshiro) [reply go.mod:65]: After doing some reading, it looks like using replace statements within the downstream module is the only way to handle cases like this, where the upstream is using replacements. 

Can you add a comment above the `github.com/docker/distribution` and `github.com/docker/docker` replacements that they are required for `github.com/deislabs/oras v0.9.0` so that their intention is clear in the future? Thanks!

Comment (jdolitsky) [reply go.mod:65]: @ikaneshiro done!

**Review by ikaneshiro — APPROVED** [review]
Lgtm

**Comment by ikaneshiro** [comment]
Looks like the CI workflow wasn't successfully triggered. I'm going to close and reopen this PR to try to kick it.
